peps/pep-0728.rst

```diff
@@ -0,0 +1,639 @@
+
+There are some advantages to this approach:
+
+- Inheritance works naturally. ``__extra__`` defined on a TypedDict will also
+  be available to its subclasses.
+
+- We can build on top of
+  :pep:`the type consistency rules defined in PEP 589 <589#type-consistency>`.
+  ``__extra__`` can be treated as a pseudo-key in terms of type consistency.
+
+- There is no need to introduce a syntax to specify the type of the extra
+  fields.
+
+- We can precisely type the extra fields without making ``__extra__`` the union
+  of known fields.
+
+Specification
+=============
+
+This specification is structured to parallel :pep:`589` to highlight changes to
+the original TypedDict specification.
+
+Extra fields are treated as non-required fields having the same type of
+``__extra__`` whose keys are allowed when determining
+:pep:`supported and unsupported operations <589>`.
+
+Using TypedDict Types
+---------------------
+
+For a TypedDict type that has the ``__extra__`` field, during construction, the
+value type of each unknown field is expected to be non-required and compatible
+with the type of ``__extra__``. For example::
+
+    class Movie(TypedDict):
+        name: str
+        __extra__: bool
+    
+    a: Movie = {"name": "Blade Runner", "novel_adaptation": True}  # OK
+    b: Movie = {
+        "name": "Blade Runner",
+        "year": 1982,  # Not OK. 'int' is incompatible with 'bool'
+    }  
+
+In this example, ``__extra__: bool`` does not mean that ``Movie`` has a required
+string key "__extra__" whose value type is ``bool``. Instead, it specifies that
+fields other than "name" have a value type of ``bool`` and are non-required.
+
+The alternative inline syntax is also supported::
+
+    Movie = TypedDict("Movie", {"name": str, "__extra__": bool})
+
+Accessing extra fields is allowed. Type checkers must infer its value type from
+the type of ``__extra__``::
+
+    def f(movie: Movie) -> None:
+        reveal_type(movie["name"])              # Revealed type is 'str'
+        reveal_type(movie["novel_adaptation"])  # Revealed type is 'bool'
+
+Interaction with PEP 705
+------------------------
+
+When ``__extra__`` is annotated with ``ReadOnly[]``, the extra fields on the
+TypedDict have the properties of read-only fields. This affects subclassing
+according to the inheritance rules specified in :pep:`PEP 705 <589#Inheritance>`.
+
+Notably, a subclass of the TypedDict type may redeclare the type of ``__extra__``
+or extra fields if the TypedDict type declares ``__extra__`` to be read-only.
+
+Interaction with Totality
+-------------------------
+
+It is an error to use ``Required[]`` or ``NotRequired[]`` with the special
+``__extra__`` field. ``total=False`` and ``total=True`` have no effect on
+``__extra__`` itself.
+
+The extra fields are non-required, regardless of the totality of the TypedDict.
+Operations that are available to ``NotRequired`` fields should also be available
+to the extra fields::
+
+    def f(movie: Movie) -> None:
```

Comment:
Might be better to repeat the definition of the `Movie` class here, since it's some way away.